lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs

```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- |
+Copyright: © 2018-2022 IOHK
+License: Apache-2.0
+
+Data type 'TxHistory' for storing a set of transactions.
+Transactions are encoded "as" expressed in DB tables.
+
+-}
+module Cardano.Wallet.DB.Store.Transactions.Model
+    ( DeltaTxHistory (..)
+    , TxHistory
+    , TxHistoryF (TxHistoryF)
+    , TxRelationF (..)
+    , tokenCollateralOrd
+    , tokenOutOrd
+    , mkTxHistory
+    , TxRelationTxOutCtxK (..)
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.DB.Sqlite.Schema
+    ( TxCollateral (..)
+    , TxCollateralOut (..)
+    , TxCollateralOutToken (..)
+    , TxIn (..)
+    , TxOut (..)
+    , TxOutToken (..)
+    , TxWithdrawal (..)
+    )
+import Cardano.Wallet.DB.Sqlite.Types
+    ( TxId (TxId) )
+import Cardano.Wallet.Primitive.Types.RewardAccount
+    ( RewardAccount )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( AssetId (AssetId) )
+import Cardano.Wallet.Primitive.Types.TokenPolicy
+    ( TokenName, TokenPolicyId )
+import Cardano.Wallet.Primitive.Types.TokenQuantity
+    ( TokenQuantity )
+import Control.Arrow
+    ( (&&&) )
+import Data.Delta
+    ( Delta (..) )
+import Data.Foldable
+    ( fold )
+import Data.Generics.Internal.VL
+    ( view, (^.) )
+import Data.List
+    ( sortOn )
+import Data.Map.Strict
+    ( Map )
+import Data.Word
+    ( Word32 )
+import Fmt
+    ( Buildable (build) )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Types.Coin as W
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.Tx as W
+import qualified Data.Map.Strict as Map
+
+-- | A context that carries a TxOut together with its tokens
+-- (this will be needed in the future for the DB Layer).
+data WithTxOut a =
+    WithTxOut
+    { withTxOut_value :: a, withTxOut_context :: Maybe (TxOut, [TxOutToken]) }
+    deriving ( Show, Eq, Functor )
+
+-- | A kind to index the 2 flavours of a 'TxRelationF', with or without 'TxOuts'
+data TxRelationTxOutCtxK
+    = NoTxOutCtx
+    | WithTxOutCtx
+
+-- | Define the TxOut context type
+type family TxRelationTxOutCtx f a where
+    TxRelationTxOutCtx 'NoTxOutCtx a = a
+    TxRelationTxOutCtx 'WithTxOutCtx a = WithTxOut a
```

Comment:
```suggestion
-- | Decorate (or not) a value with a 'TxOut'.
type family DecorateTxOut d a where
    DecorateTxOut 'Without a = a
    DecorateTxOut 'With a = WithTxOut a
```

Reply:
Another possible renaming could be `TxOutDecoration 'With` or `TxOutDecoration 'Without`

Reply:
Ah. I like to focus on the decoration that is added (`TxOut`) rather than the thing that is decorated (`TxIn` or collateral input) — If `TxOut` is part of the type name, then this is more indicative of what data the type contains / adds.